### PR TITLE
Created ad hoc operations cucumber tests.

### PR DESCRIPTION
**Pull request by carlogeenen**

### CODE REVIEW

**Comment by Sander3003** [comment]
Can you add the use of cucumber to the documentation as well? http://54.77.62.182/documentation/chapter3/section1.html

File: integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        final TestCaseResult runTestStepByName = this.testCaseRunner.runWsdlTestCase(this.testCase, PROPERTIES_MAP,
+                TEST_CASE_NAME_REQUEST);
+
+        final TestStepResult runTestStepByNameResult = runTestStepByName.getRunTestStepByName();
+        final WsdlTestCaseRunner wsdlTestCaseRunner = runTestStepByName.getResults();
+        assertEquals(TestStepStatus.OK, runTestStepByNameResult.getStatus());
+
+        for (final TestStepResult tcr : wsdlTestCaseRunner.getResults()) {
+            this.response = ((MessageExchange) tcr).getResponseContent();
+            this.correlationUidMatcher = this.correlationUidPattern.matcher(this.response);
```
Comment (eduarddrenth) [inline integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java]: You loop over all results, set the matcher each time and use the matcher set in the last iteration. Is that what is intended here? If you realy need the last result why loop?

File: integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/support/TestCaseRunner.java [hunk 2]
```diff
@@ -36,7 +36,7 @@ public TestCaseResult runWsdlTestCase(final TestCase testCase, final Map<String,
         final WsdlTestCaseRunner wsdlTestCaseRunner = new WsdlTestCaseRunner(wsdlTestCase, new PropertiesMap());
 
         try {
-            Thread.sleep(30000);
+            Thread.sleep(25000);
         } catch (final InterruptedException ex) {
             Thread.currentThread().interrupt();
```
Comment (eduarddrenth) [inline integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/support/TestCaseRunner.java]: Thread has already been interrupted, if desired you can log that interrupt (the exception here), normally you would just for example leave the method here, you can also just throw the exception.

File: integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java (hunk 1 above)
Comment (eduarddrenth) [inline integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java]: You loop over all results, set the matcher each time and use the matcher set in the last iteration. Is that what is intended here? If you realy need the last result why loop?

File: integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        }
+        this.correlationUidMatcher.find();
```
Comment (eduarddrenth) [inline integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java]: What if nothing is found?

File: integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    private static final String PATH_RESULT_VERSION = "/Envelope/Body/GetAssociationLnObjectsResponse/AssociationLnList/AssociationLnListElement/Version/text()";
+    private static final String PATH_RESULT_LOGICALNAME = "/Envelope/Body/GetAssociationLnObjectsResponse/AssociationLnList/AssociationLnListElement/LogicalName/text()";
+    private static final String PATH_RESULT_ATTRIBUTE_ID = "/Envelope/Body/GetAssociationLnObjectsResponse/AssociationLnList/AssociationLnListElement/AccessRights/AttributeAccess/AttributeAccessItem/AttributeId/text()";
+    private static final String PATH_RESULT_ACCESS_MODE = "/Envelope/Body/GetAssociationLnObjectsResponse/AssociationLnList/AssociationLnListElement/AccessRights/AttributeAccess/AttributeAccessItem/AccessMode/text()";
+
+    private static final String XPATH_MATCHER_RESULT = "OK";
+    private static final String XPATH_MATCHER_RESULT_DECIMAL = "\\d+";
+    private static final String XPATH_MATCHER_RESULT_ACCESS_MODE = "\\w+\\_\\w+";
+
+    private static final String SOAP_PROJECT_XML = "../cucumber/soap-ui-project/FLEX-OVL-V3---SmartMetering-soapui-project.xml";
```
Comment (robindenadel) [inline integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java]: This constant is in all the classes in the package com.alliander.osgp.platform.cucumber.smartmeteringmonitoring as well. Create a super_super class where this constant is in; and a super class per wsdl where constants like TEST_SUITE_XML are in

File: integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    private static final String TEST_CASE_NAME_RESPONSE = "GetGetAssociationLnObjectsResponse - Request 1";
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AssociationLnObjects.class);
+    private static final Map<String, String> PROPERTIES_MAP = new HashMap<>();
+
+    private Pattern correlationUidPattern;
+    private Matcher correlationUidMatcher;
+
+    @Autowired
+    private WsdlProjectFactory wsdlProjectFactory;
```
Comment (robindenadel) [inline integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java]: make this property protected and move it to the super_super

File: integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+    @Autowired
+    private TestCaseRunner testCaseRunner;
```
Comment (robindenadel) [inline integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java]: make this property protected and move it to the super_super

File: integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+    @Autowired
+    private RunXpathResult runXpathResult;
```
Comment (robindenadel) [inline integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java]: make this property protected and move it to the super_super

File: integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java (hunk 4 above)
Comment (robindenadel) [inline integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java]: refactor the existing classes according to this

File: integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java (hunk 5 above)
Comment (robindenadel) [inline integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java]: refactor the existing classes according to this

File: integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java (hunk 6 above)
Comment (robindenadel) [inline integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java]: refactor the existing classes according to this

File: integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java (hunk 7 above)
Comment (robindenadel) [inline integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java]: refactor the existing classes according to this

File: integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+    @Then("^the objectlist should be returned$")
+    public void theObjectlistShouldBeReturned() throws Throwable {
+        PROPERTIES_MAP.put("CorrelationUid", this.correlationUid);
+
+        final TestCaseResult runTestStepByName = this.testCaseRunner.runWsdlTestCase(this.testCase, PROPERTIES_MAP,
+                TEST_CASE_NAME_RESPONSE);
+        final TestStepResult runTestStepByNameResult = runTestStepByName.getRunTestStepByName();
+        final WsdlTestCaseRunner wsdlTestCaseRunner = runTestStepByName.getResults();
+        assertEquals(TestStepStatus.OK, runTestStepByNameResult.getStatus());
```
Comment (robindenadel) [inline integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java]: This code block is in a lot of other classes; extract a method, move it to the super_super. Refactor existing classes.

File: integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+        for (final TestStepResult tcr : wsdlTestCaseRunner.getResults()) {
+            LOGGER.info(TEST_CASE_NAME_RESPONSE + " response {}",
+                    this.response = ((MessageExchange) tcr).getResponseContent());
+        }
```
Comment (robindenadel) [inline integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java]: seems like this loop is part of the code block that is in all the test classes

File: integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+    @Autowired
+    private DeviceId deviceId;
+
+    @Autowired
+    private OrganisationId organisationId;
+
+    @When("^the retrieve association LN objectlist request is received$")
+    public void theRetrieveAssociationLNObjectlistRequestIsReceived() throws Throwable {
+        this.correlationUidPattern = Pattern.compile(this.organisationId.getOrganisationId()
+                + "\\|\\|\\|\\S{17}\\|\\|\\|\\S{17}");
```
Comment (robindenadel) [inline integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java]: "\\|\\|\\|\\S{17}\\|\\|\\|\\S{17}" should be a constant in the super_super. Refactor other classes

File: integration-tests-cucumber/src/test/resources/AssociationLnObjects.feature [hunk 11]
```diff
@@ -0,0 +1,11 @@
+Feature:
```
Comment (robindenadel) [inline integration-tests-cucumber/src/test/resources/AssociationLnObjects.feature]: if we organise the scenario's per wsdl, i think it is easier to get an overview.
In that case, the first two lines of each scenario can be moved to Background

File: integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+import cucumber.api.java.en.And;
+import cucumber.api.java.en.Then;
+import cucumber.api.java.en.When;
+
+public class AssociationLnObjects {
+    private TestCase testCase;
+    private String response;
+    private String correlationUid;
+
+    private static final String PATH_RESULT = "/Envelope/Body/GetAssociationLnObjectsResponse/Result/text()";
```
Comment (eduarddrenth) [inline integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/AssociationLnObjects.java]: reminder / warning the xpath text() functions selects ALL childnodes of type text under the Result node. So you may get more text then you want

File: integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/ConfigurationObjects.java [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+
+        final TestStepResult runTestStepByNameResult = runTestStepByName.getRunTestStepByName();
+        final WsdlTestCaseRunner wsdlTestCaseRunner = runTestStepByName.getResults();
+        assertEquals(TestStepStatus.OK, runTestStepByNameResult.getStatus());
+
+        for (final TestStepResult tcr : wsdlTestCaseRunner.getResults()) {
+            this.response = ((MessageExchange) tcr).getResponseContent();
+            this.correlationUidMatcher = this.correlationUidPattern.matcher(this.response);
+        }
+        this.correlationUidMatcher.find();
```
Comment (eduarddrenth) [inline integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/ConfigurationObjects.java]: asserttrue....

File: integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/SpecificConfigurationObject.java [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+
+        final TestStepResult runTestStepByNameResult = runTestStepByName.getRunTestStepByName();
+        final WsdlTestCaseRunner wsdlTestCaseRunner = runTestStepByName.getResults();
+        assertEquals(TestStepStatus.OK, runTestStepByNameResult.getStatus());
+
+        for (final TestStepResult tcr : wsdlTestCaseRunner.getResults()) {
+            this.response = ((MessageExchange) tcr).getResponseContent();
+            this.correlationUidMatcher = this.correlationUidPattern.matcher(this.response);
+        }
+        this.correlationUidMatcher.find();
```
Comment (eduarddrenth) [inline integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/SpecificConfigurationObject.java]: asserttrue

File: integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/SynchronizeTime.java [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+        final TestStepResult runTestStepByNameResult = runTestStepByName.getRunTestStepByName();
+        final WsdlTestCaseRunner wsdlTestCaseRunner = runTestStepByName.getResults();
+        assertEquals(TestStepStatus.OK, runTestStepByNameResult.getStatus());
+
+        for (final TestStepResult tcr : wsdlTestCaseRunner.getResults()) {
+            this.response = ((MessageExchange) tcr).getResponseContent();
+            this.correlationUidMatcher = this.correlationUidPattern.matcher(this.response);
+        }
+        this.correlationUidMatcher.find();
```
Comment (eduarddrenth) [inline integration-tests-cucumber/src/test/java/com/alliander/osgp/platform/cucumber/smartmeteringadhoc/SynchronizeTime.java]: asserttrue

**Pull request merged**